C++20 std::coroutine cmake detection

Configuring HPX head on a box with GNU g++ 10.3.0, I am trying to get
```
-- Performing Test HPX_WITH_CXX20_COROUTINES - Failed
```
to pass, but am getting lost in the cmake files.

According to g++'s error message, it seems that a `promise_type` also needs to define `unhandled_exception`, so I applied
```
index 619fb73205..bba9190fac 100644
--- a/cmake/tests/cxx20_coroutines.cpp
+++ b/cmake/tests/cxx20_coroutines.cpp
@@ -30,6 +30,7 @@ struct resumable
         }
 
         void return_value(int value) const noexcept {}
+        void unhandled_exception() const noexcept {}
     };
 
     bool await_ready() const noexcept
```

Copying the patched `cmake/tests/cxx20_coroutines.cpp` to another directory, and adding
```
$ cat CMakeLists.txt 
cmake_minimum_required(VERSION 3.00)
project(coro)
set(CMAKE_CXX_STANDARD 20)
set(CMAKE_CXX_STANDARD_REQUIRED True)
add_executable(coro cxx20_coroutines.cpp)
target_compile_options(coro PUBLIC "-fcoroutines")
```
showed that it compiles successfully.

I tried configuring hpx with
```
cmake \
-DCMAKE_BUILD_TYPE=Release \
-DHPX_WITH_DOCUMENTATION=OFf \
-DHPX_WITH_EXAMPLES=OFF \
-DHPX_WITH_TESTS=OFF \
-DHPX_WITH_NETWORKING=OFF \
-DHPX_WITH_DISTRIBUTED_RUNTIME=OFF \
-DCMAKE_CXX_STANDARD=20 -DHPX_USE_CMAKE_CXX_STANDARD=ON \
..
```
and
```
index fd7ab0dad2..3b5a79bcaf 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1623,6 +1623,8 @@ if(CMAKE_COMPILER_IS_GNUCXX AND HPX_WITH_GCC_VERSION_CHECK)
     hpx_error(
       "GCC 7.0 or higher is required. Specify HPX_WITH_GCC_VERSION_CHECK=OFF to ignore this error."
     )
+  elif(CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 10.0)
+      hpx_add_target_compile_option(-fcoroutines PUBLIC)
   endif()
 endif()
 
```
but that didn't do the trick...

@msimberg I'm puzzled that it worked for you after just adding `unhandled_exception`: GCC 10 implements std::coroutine, but requires -fcoroutines to compile. In the outside-of-hpx CMakeLists.txt I posted above, the compilation is not successful without the
```
target_compile_options(coro PUBLIC "-fcoroutines")
```
line. I just checked the [GCC 11 release notes](https://gcc.gnu.org/gcc-11/changes.html#cxx), and coroutines aren't mentioned, so I assume no change, and the flag is still necessary?

A `git grep fcoroutine` in hpx and cmake gets no hits (apart from the version with my attempt at adding the flag above), so the question is:

How does -fcoroutines get added for you?


I had a go with
```
debug trycompile on
...
-- Performing Test HPX_WITH_CXX20_COROUTINES - Failed
CMake Error at cmake/HPX_PerformCxxFeatureTests.cmake:63 (message):
  stopping
```
but haven't found the failure reason, e.g., coroutine doesn't appear in `obj/CMakeFiles/CMakeOutput.log`.
I do see
```
$ ls -1 obj/CMakeFiles/config_tests/CMakeFiles/CMakeTmp/CMakeFiles/cmTC_7d674.dir
DependInfo.cmake
build.make
cmake_clean.cmake
compiler_depend.make
compiler_depend.ts
depend.make
flags.make
link.txt
progress.make
```
and note
```
$ cat flags.make 
# CMAKE generated file: DO NOT EDIT!
# Generated by "Unix Makefiles" Generator, CMake Version 3.21

# compile CXX with /usr/bin/c++
CXX_DEFINES = 

CXX_INCLUDES = 

CXX_FLAGS =  -Werror  -std=c++2a
```

Ah!
```
$ cd obj/CMakeFiles/config_tests/CMakeFiles/CMakeTmp
$ make cmTC_7d674
make  -f CMakeFiles/Makefile2 cmTC_7d674
/usr/pkg/bin/cmake -E cmake_progress_start /usr/src/local/hpxrestart/hpx/obj/CMakeFiles/config_tests/CMakeFiles/CMakeTmp/CMakeFiles 2
make  -f CMakeFiles/Makefile2 CMakeFiles/cmTC_7d674.dir/all
make  -f CMakeFiles/cmTC_7d674.dir/build.make CMakeFiles/cmTC_7d674.dir/depend
cd /usr/src/local/hpxrestart/hpx/obj/CMakeFiles/config_tests/CMakeFiles/CMakeTmp && /usr/pkg/bin/cmake -E cmake_depends "Unix Makefiles" /usr/src/local/hpxrestart/hpx/obj/CMakeFiles/config_tests/CMakeFiles/CMakeTmp /usr/src/local/hpxrestart/hpx/obj/CMakeFiles/config_tests/CMakeFiles/CMakeTmp /usr/src/local/hpxrestart/hpx/obj/CMakeFiles/config_tests/CMakeFiles/CMakeTmp /usr/src/local/hpxrestart/hpx/obj/CMakeFiles/config_tests/CMakeFiles/CMakeTmp /usr/src/local/hpxrestart/hpx/obj/CMakeFiles/config_tests/CMakeFiles/CMakeTmp/CMakeFiles/cmTC_7d674.dir/DependInfo.cmake
Scanning dependencies of target cmTC_7d674
make  -f CMakeFiles/cmTC_7d674.dir/build.make CMakeFiles/cmTC_7d674.dir/build
[ 50%] Building CXX object CMakeFiles/cmTC_7d674.dir/cxx20_coroutines.cpp.o
/usr/bin/c++   -Werror  -std=c++2a -o CMakeFiles/cmTC_7d674.dir/cxx20_coroutines.cpp.o -c /usr/src/local/hpxrestart/hpx/cmake/tests/cxx20_coroutines.cpp
In file included from /usr/src/local/hpxrestart/hpx/cmake/tests/cxx20_coroutines.cpp:8:
/usr/include/g++/coroutine:295:2: error: #error "the coroutine header requires -fcoroutines"
  295 | #error "the coroutine header requires -fcoroutines"
      |  ^~~~~
```
so yes, it is back to the question of how to shoe horn "-fcoroutines" in...

Amazing: so you found out that GCC 11 no longer uses -fcoroutines, and didn't declare this improvement in the release notes.

That says that my first attempt at a patch in the original bug report has to be `ver >=10.0 && ver < 11`.

It still leaves the problem of shoe horning the `-fcoroutines` flag in just for g++ 10, which is the only situation for which that flag is both necessary and valid. It turns out that my  first`hpx_add_target_compile_option` guess above will not add the flag for test targets. I need to set `CMAKE_CXX_FLAGS` for it to be set everywhere (`add_compile_options` similarly insufficient).

This version works:
```
diff --git a/CMakeLists.txt b/CMakeLists.txt
index fd7ab0dad2..5ff09d0974 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1626,6 +1626,11 @@ if(CMAKE_COMPILER_IS_GNUCXX AND HPX_WITH_GCC_VERSION_CHECK)
   endif()
 endif()
 
+if(CMAKE_COMPILER_IS_GNUCXX AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 10 AND CMAKE_CXX_COMPILER_VERSION VERSION_LESS 11)
+  SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fcoroutines")
+  hpx_info("Detected g++ ${CMAKE_CXX_COMPILER_VERSION} and added -fcoroutines flag")
+endif()
+
 if(MSVC)
   if(MSVC_VERSION LESS 1900)
     hpx_error("MSVC x64 2015 or higher is required.")
@@ -1768,7 +1773,8 @@ if(NOT WIN32)
     target_link_libraries(hpx_base_libraries INTERFACE supc++)
   endif()
 
-  if(NOT ${CMAKE_SYSTEM_NAME} MATCHES "FreeBSD")
+  find_library(DL_LIBRARY NAMES dl)
+  if(DL_LIBRARY)
     target_link_libraries(hpx_base_libraries INTERFACE dl)
   endif()
 
```
resulting in
```
-- Detected g++ 10.3.0 and added -fcoroutines flag
...
-- Performing Test HPX_WITH_CXX20_COROUTINES - Success
```

BTW I went down this -fcoroutines path, as I was hoping it might avoid
```
/usr/src/local/hpxrestart/hpx/libs/core/coroutines/include/hpx/coroutines/detail/context_impl.hpp:145:2: error: #error No default_context_impl available for this system
```
by moving portability issues to the compiler from the hpx code.

If my compiler implements std::coroutine, why is `hpx/coroutines/detail/context_impl.hpp` complaining? What do I need to implement?

--- a/libs/core/futures/include/hpx/futures/traits/detail/future_await_traits.hpp
+++ b/libs/core/futures/include/hpx/futures/traits/detail/future_await_traits.hpp
@@ -258,6 +258,11 @@ namespace std {
                     this->base_type::set_value(std::forward<U>(value));
                 }
 
+                void unhandled_exception() noexcept
+                {
+                    this->base_type::set_exception(std::current_exception());
+                }
+
                 HPX_NODISCARD HPX_FORCEINLINE static void* operator new(
                     std::size_t size)
                 {
@@ -299,10 +304,15 @@ namespace std {
                     this->base_type::set_value();
                 }
 
+                void unhandled_exception() noexcept
+                {
+                    this->base_type::set_exception(std::current_exception());
+                }
+
                 HPX_NODISCARD HPX_FORCEINLINE static void* operator new(
                     std::size_t size)
                 {
-                    return base_type::allocate();
+                    return base_type::allocate(size);
                 }
 
                 HPX_FORCEINLINE static void operator delete(
@@ -341,6 +351,11 @@ namespace std {
                 void return_value(U&& value)
                 {
                     this->base_type::set_value(std::forward<U>(value));
+                }
+
+                void unhandled_exception() noexcept
+                {
+                    this->base_type::set_exception(std::current_exception());
                 }
 
                 HPX_NODISCARD HPX_FORCEINLINE static void* operator new(
@@ -384,10 +399,15 @@ namespace std {
                     this->base_type::set_value();
                 }
 
+                void unhandled_exception() noexcept
+                {
+                    this->base_type::set_exception(std::current_exception());
+                }
+
                 HPX_NODISCARD HPX_FORCEINLINE static void* operator new(
                     std::size_t size)
                 {
-                    return base_type::allocate();
+                    return base_type::allocate(size);
                 }
 
                 HPX_FORCEINLINE static void operator delete(

